README.md
# <img src="https://docs.lido.fi/img/logo.svg" alt="Lido" width="46"/> Lido Depositor bot

## Description

Depositor and pauser bots are parts of [Deposit Security Module](https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-5.md#mitigations-for-deposit-front-running-vulnerability).

**The Depositor Bot** obtains signed deposit messages from Council Daemons. 
Once a sufficient number of messages is collected to constitute a quorum, the bot proceeds to initiate a deposit into the designated staking module. 
This deposit is executed using the depositBufferedEther function within the "DepositSecurityModule" smart contract.

**The Pauser Bot** obtains pause message from Council Daemon and enacts pause on the specified staking module.

## Table of Contents

- [Running Daemon](#running-daemon)
- [Variables](#variables)
  - [Required variables](#required-variables)
  - [Additional variables](#additional-variables)
- [Metrics and logs](#metrics-and-logs)
- [Development](#development)
  - [Install](#install)
  - [Tests](#tests)
  - [Release flow](#release-flow)
- [Annotations to code](#annotations-to-code)

## Running Daemon

1. Create `.env` file 
2. Setup variables
   - Set WEB3_RPC_ENDPOINTS
   - Set WALLET_PRIVATE_KEY
   - Set CREATE_TRANSACTIONS to true
   - Set MESSAGE_TRANSPORTS to rabbit
   - Set RABBIT_MQ_URL, RABBIT_MQ_USERNAME and RABBIT_MQ_PASSWORD
3. ```docker-compose up```
4. Send metrics and logs to grafana
5. Setup alerts

## Variables

### Required variables

| Variable                  | Default                                    | Description                                                                                                              |
|---------------------------|--------------------------------------------|--------------------------------------------------------------------------------------------------------------------------|
| WEB3_RPC_ENDPOINTS        | -                                          | List of rpc endpoints that will be used to send requests comma separated (`,`)                                           |
| WALLET_PRIVATE_KEY        | -                                          | Account private key                                                                                                      |
| CREATE_TRANSACTIONS       | false                                      | If true then tx will be send to blockchain                                                                               |
| LIDO_LOCATOR              | 0xC1d0b3DE6792Bf6b4b37EccdcC24e45978Cfd2Eb | Lido Locator address. Mainnet by default. Other networks could be found [here](https://docs.lido.fi/deployed-contracts/) |
| DEPOSIT_CONTRACT          | 0x00000000219ab540356cBB839Cbe05303d7705Fa | Ethereum deposit contract address                                                                                        |
| DEPOSIT_MODULES_WHITELIST | 1                                          | List of staking module's ids in which the depositor bot will make deposits                                               |
| ---                       | ---	                                       | ---                                                                                                                      |
| MESSAGE_TRANSPORTS        | -                                          | Transports used in bot. One of/or both: rabbit/kafka                                                                     |
| RABBIT_MQ_URL             | -                                          | RabbitMQ url                                                                                                             |
| RABBIT_MQ_USERNAME        | -                                          | RabbitMQ username for virtualhost                                                                                        |
| RABBIT_MQ_PASSWORD        | -                                          | RabbitMQ password for virtualhost                                                                                        |
| ---                       | --- _kafka is not used at the moment_ ---  | ---                                                                                                                      |
| KAFKA_BROKER_ADDRESS_1    | -                                          | Kafka servers url and port                                                                                               |
| KAFKA_USERNAME            | -                                          | Kafka username                                                                                                           |
| KAFKA_PASSWORD            | -                                          | Password for kafka                                                                                                       |
| KAFKA_NETWORK             | -                                          | Network type (mainnet or goerli)                                                                                         |
| KAFKA_TOPIC               | -                                          | Kafka topic name (for msg receiving)                                                                                     |
| KAFKA_GROUP_PREFIX        | -                                          | Just for staging (staging-)                                                                                              |

### Additional variables

| Variable                          | Default    | Description                                                                                                              |
|-----------------------------------|------------|--------------------------------------------------------------------------------------------------------------------------|
| MIN_PRIORITY_FEE                  | 50 mwei    | Min priority fee that will be used in tx                                                                                 |
| MAX_PRIORITY_FEE                  | 10 gwei    | Max priority fee that will be used in tx                                                                                 |
| MAX_GAS_FEE                       | 100 gwei   | Bot will wait for a lower price. Treshold for gas_fee                                                                    |
| CONTRACT_GAS_LIMIT                | 15000000   | Default transaction gas limit                                                                                            |
| AUCTION_BUNDLER_URIS              | -          | List of relays uris                                                                                                      |
| AUCTION_BUNDLER_PRIVATE_KEY       | -          | Private key - Used to identify account for relays (should NOT be equal to WALLET private key)                            |
| GAS_FEE_PERCENTILE_1              | 20         | Percentile for first recommended fee calculation                                                                         |
| GAS_FEE_PERCENTILE_DAYS_HISTORY_1 | 1          | Percentile for first recommended calculates from N days of the fee history                                               |
| GAS_PRIORITY_FEE_PERCENTILE       | 25         | Priority transaction will be N percentile from priority fees in last block (min MIN_PRIORITY_FEE - max MAX_PRIORITY_FEE) |
| MAX_BUFFERED_ETHERS               | 5000 ether | Maximum amount of ETH in the buffer, after which the bot deposits at any gas                                             |
| PROMETHEUS_PORT                   | 9000       | Port with metrics server                                                                                                 |
| PULSE_SERVER_PORT                 | 9010       | Port with bot`s status server                                                                                            |
| MAX_CYCLE_LIFETIME_IN_SECONDS     | 1200       | Max lifetime of usual cycle. If cycle will not end in this time, bot will crush                                          |

## Metrics and logs

Metrics list could be found in [source code](src/metrics/metrics.py).
Prometheus server hosted on `http://localhost:${{PROMETHEUS_PORT}}/`.

## Development

### Install

```bash
git clone git@github.com:lidofinance/depositor-bot.git
cd depositor-bot
poetry install
```

To run bot

```bash
# For depositor bot
poetry run python src/depositor.py

# For pause bot
poetry run python src/pauser.py
```

### Tests

#### Run unit tests

```bash
poetry run pytest tests -m unit
```

#### Run integration tests.

Install Hardhat and run goerli fork
```bash
npm install --save-dev hardhat
npx hardhat node --fork ${{ WEB3_RPC_ENDPOINT }} &
```

```bash
poetry run pytest tests -m integration
```

### Release flow

To create a new release:

1. Merge all changes to the `main` branch.
1. After the merge, the `Prepare release draft` action will run automatically. When the action is complete, a release draft is created.
1. When you need to release, go to Repo → Releases.
1. Publish the desired release draft manually by clicking the edit button - this release is now the `Latest Published`.
1. After publication, the action to create a release bump will be triggered automatically.

## Annotations to code

Council daemons sends messages to depositor and pauser bot throw kafka or rabbit.

![img.png](static/img.png)

Every 5 blocks Executor instance calls depositor bot. If Depositor returns failed status, executor executes depositor with next block instead of waiting for 5 blocks. 

![img.png](static/img1.png)

Depositor bot makes deposit only if next conditions are ok
- Gas fee is low enough
- There are enough buffered eth ready for deposit
- There are enough messages from Council Daemons to form a quorum
- Module is active and not paused

![img.png](static/img2.png)

Message filtering process
1. Filter all invalid messages. (Invalid schema or signature)
2. Filter all expired messages
3. Filter all messages with block number less or equal to latest block in depositor bot and with outdated nonce or deposit root.

![img.png](static/img3.png)


CHANGELOG.md
# This file is no longer used by semantic-release
For the latest release notes, please see the [GitHub releases page](https://github.com/lidofinance/depositor-bot/releases).

## [4.2.1](https://github.com/lidofinance/depositor-bot/compare/4.2.0...4.2.1) (2023-05-15)

# [4.2.0](https://github.com/lidofinance/depositor-bot/compare/v3.3.1...4.2.0) (2023-04-21)

### Bug Fixes

* _do_classic_deposit ([4d6d379](https://github.com/lidofinance/depositor-bot/commit/4d6d379d8bc6283ac02c68361c9fe1aeac705e39))
* addresses ([228e5a2](https://github.com/lidofinance/depositor-bot/commit/228e5a2c93a400b470e86ce980cdeb99096eb300))
* BOTH -> ALL ([590e006](https://github.com/lidofinance/depositor-bot/commit/590e006abbf9ce5a49a07d28eeea1c7ff166efe4))
* bytearray ([62182da](https://github.com/lidofinance/depositor-bot/commit/62182da9487d8ab14c5471b340829bb3d9020943))
* curl dockerfile ([65b70ec](https://github.com/lidofinance/depositor-bot/commit/65b70ecfb6a132d88b6abe8bc2c8fd99b034e9b6))
* curl version ([e8f0e90](https://github.com/lidofinance/depositor-bot/commit/e8f0e908ebc5499b271c9fc5baf8f137fcfda8b6))
* disable _available_keys_issue ([1066fe2](https://github.com/lidofinance/depositor-bot/commit/1066fe2b5393a6104c73925c5bd1277f452e4a89))
* dockerfile ([334b7e5](https://github.com/lidofinance/depositor-bot/commit/334b7e52a4a688e31d64bf466053dce7080aafc8))
* env ([01f7884](https://github.com/lidofinance/depositor-bot/commit/01f78844570a6cdb55e72efc5b41cfdf720d09bb))
* env variables ([48a952f](https://github.com/lidofinance/depositor-bot/commit/48a952f220d45b4cf2e10b17f95f22e4d0f6f441))
* eof ([9becbdb](https://github.com/lidofinance/depositor-bot/commit/9becbdbaea88a3a63fd406a2fd2a69e9b73d68bc))
* eof ([ba821ad](https://github.com/lidofinance/depositor-bot/commit/ba821ad112e5ca68427c90afa8e224f004669133))
* eof ([f6effed](https://github.com/lidofinance/depositor-bot/commit/f6effedd61658f15a7fb5c46624fb797d962784b))
* eof ([cbc59a2](https://github.com/lidofinance/depositor-bot/commit/cbc59a20c5ecf653524c17f8393d2e0c74cdc86e))
* exception format ([da65e07](https://github.com/lidofinance/depositor-bot/commit/da65e079f36eb9565472b78c0f9459fc991d5c96))
* format ([2771bd9](https://github.com/lidofinance/depositor-bot/commit/2771bd94a05a8f137f7b8d6a011683624808f46a))
* get nonce ([d89a9be](https://github.com/lidofinance/depositor-bot/commit/d89a9bef2f11c6461aa421fac0bdd05876a118dd))
* get_deposit_messages_sign_filter and logs ([d2096a3](https://github.com/lidofinance/depositor-bot/commit/d2096a33ff55ef885ba2fd8a0b0f336f226ad57b))
* git version ([9020dda](https://github.com/lidofinance/depositor-bot/commit/9020dda12bad31d4fc047c9c0b5b67c78b3446fe))
* goerli addresses ([2b9acc3](https://github.com/lidofinance/depositor-bot/commit/2b9acc31c57dc2067c34bff8c9870d7585cf50de))
* goerli contracts ([8282b0f](https://github.com/lidofinance/depositor-bot/commit/8282b0f1764d158927b74766818fb5f9ae08ece6))
* is paused ([b8d96d4](https://github.com/lidofinance/depositor-bot/commit/b8d96d43ec2a3baa46d92b27e91308a853b40639))
* is_paused -> is_active ([d1111f8](https://github.com/lidofinance/depositor-bot/commit/d1111f846d39f0454f80e2487884c3d6c716800b))
* len ([5141462](https://github.com/lidofinance/depositor-bot/commit/5141462161a8b064cd915ec3aeb074f04a04bb28))
* new contract addresses ([858d4fb](https://github.com/lidofinance/depositor-bot/commit/858d4fba1a5537dbac9dc5aabbeae2a36c53ab7c))
* pause bot ([53bb0d7](https://github.com/lidofinance/depositor-bot/commit/53bb0d71ab728e8adea4776d289ab66054e450c5))
* pauser blockchain module import ([3b8f7b1](https://github.com/lidofinance/depositor-bot/commit/3b8f7b146b26641a3804e40052059b9c292b82b2))
* pauser start ([f629158](https://github.com/lidofinance/depositor-bot/commit/f629158797513016a93eb708f6c3f1210fe37abc))
* remove package ([1460b19](https://github.com/lidofinance/depositor-bot/commit/1460b19ae79168176763ce79c98615cd5cfab7f2))
* STAKING_ROUTER 0x0 address ([e78345f](https://github.com/lidofinance/depositor-bot/commit/e78345fccd677d8678c9ea90f3137282eb6767cd))
* temp fix module id ([72b4557](https://github.com/lidofinance/depositor-bot/commit/72b4557364635309b1775815fd0c1aaa4562d145))
* tests ([bc2faa1](https://github.com/lidofinance/depositor-bot/commit/bc2faa1797d390e5bfa69b3c7cdb4bbc55b5520f))
* traceback ([43ed483](https://github.com/lidofinance/depositor-bot/commit/43ed4839e9a521faabe4069b929540ba28ddff40))
* transports config ([0c1cfb6](https://github.com/lidofinance/depositor-bot/commit/0c1cfb63197c85cec65b74fb3a28de448c19e34a))
* variables ([39b168d](https://github.com/lidofinance/depositor-bot/commit/39b168d2322fc4013d9e5969d09c5ee4159260ce))
* zhejiang addresses ([b448a7e](https://github.com/lidofinance/depositor-bot/commit/b448a7eb8068ac8a3c0b9defe1d39c8d17175557))


### Features

* add Zhejiang support ([0167609](https://github.com/lidofinance/depositor-bot/commit/016760954e00b52d90ea6916f7f8abe097dbcc36))
* add Zhejiang support after re-deploy ([262be1f](https://github.com/lidofinance/depositor-bot/commit/262be1f3050b8c7589ddd84be5f9c5629bd62eb5))
* message transports config ([9a96e0d](https://github.com/lidofinance/depositor-bot/commit/9a96e0de14fda81b95dc6cfcba671a1976dafe9b))
* sr deploy ([84d5d64](https://github.com/lidofinance/depositor-bot/commit/84d5d6436abe087aeee38e296f9913040ec95642))
* staking router support ([c31ace2](https://github.com/lidofinance/depositor-bot/commit/c31ace2707286626fb32abe1c166761e7cccbebd))
* sync abi script ([cd49ab7](https://github.com/lidofinance/depositor-bot/commit/cd49ab70dcee580be8689c494fa6a79622e9ecc0))



## [3.3.1](https://github.com/lidofinance/depositor-bot/compare/v3.3.0...v3.3.1) (2022-08-09)



# [3.3.0](https://github.com/lidofinance/depositor-bot/compare/v3.2.0...v3.3.0) (2022-08-03)


### Bug Fixes

* testnet deploy ([c5ee05b](https://github.com/lidofinance/depositor-bot/commit/c5ee05b59a1d33b954988185f9bce480a768ee52))


### Features


src/variables.py
import logging
import os

from eth_account import Account
from web3 import Web3

logger = logging.getLogger(__name__)


# EL node
WEB3_RPC_ENDPOINTS = os.getenv('WEB3_RPC_ENDPOINTS', '').split(',')

# Account private key
WALLET_PRIVATE_KEY = os.getenv('WALLET_PRIVATE_KEY', None)

if WALLET_PRIVATE_KEY:
    ACCOUNT = Account.from_key(WALLET_PRIVATE_KEY)
    logger.info({'msg': 'Load account from private key.', 'value': ACCOUNT.address})
else:
    ACCOUNT = None
    logger.warning({'msg': 'Account not provided. Run in dry mode.'})

# App specific
# LIDO_LOCATOR ADDRESS
# Mainnet: 0xC1d0b3DE6792Bf6b4b37EccdcC24e45978Cfd2Eb
# Görli: 0x1eDf09b5023DC86737b59dE68a8130De878984f5
LIDO_LOCATOR = os.getenv('LIDO_LOCATOR', '0xC1d0b3DE6792Bf6b4b37EccdcC24e45978Cfd2Eb')

# DEPOSIT_CONTRACT ADDRESS
# Mainnet: 0x00000000219ab540356cBB839Cbe05303d7705Fa
# Görli: 0xff50ed3d0ec03aC01D4C79aAd74928BFF48a7b2b
DEPOSIT_CONTRACT = os.getenv('DEPOSIT_CONTRACT', '0x00000000219ab540356cBB839Cbe05303d7705Fa')

# rabbit / kafka / rabbit,kafka
MESSAGE_TRANSPORTS = os.getenv('MESSAGE_TRANSPORTS', '').split(',')

# Kafka secrets
KAFKA_BROKER_ADDRESS_1 = os.getenv('KAFKA_BROKER_ADDRESS_1')
KAFKA_USERNAME = os.getenv('KAFKA_USERNAME')
KAFKA_PASSWORD = os.getenv('KAFKA_PASSWORD')
KAFKA_NETWORK = os.getenv('KAFKA_NETWORK', 'mainnet')  # or goerli
KAFKA_TOPIC = os.getenv('KAFKA_TOPIC')
KAFKA_GROUP_PREFIX = os.getenv('KAFKA_GROUP_PREFIX', '')

# rabbit secrets
RABBIT_MQ_URL = os.getenv('RABBIT_MQ_URL', 'ws://127.0.0.1:15674/ws')
RABBIT_MQ_USERNAME = os.getenv('RABBIT_MQ_USERNAME', 'guest')
RABBIT_MQ_PASSWORD = os.getenv('RABBIT_MQ_PASSWORD', 'guest')

# Transactions settings
CREATE_TRANSACTIONS = os.getenv('CREATE_TRANSACTIONS') == 'true'

MIN_PRIORITY_FEE = Web3.to_wei(*os.getenv('MIN_PRIORITY_FEE', '50 mwei').split(' '))
MAX_PRIORITY_FEE = Web3.to_wei(*os.getenv('MAX_PRIORITY_FEE', '10 gwei').split(' '))

MAX_GAS_FEE = Web3.to_wei(*os.getenv('MAX_GAS_FEE', '100 gwei').split(' '))
CONTRACT_GAS_LIMIT = int(os.getenv('CONTRACT_GAS_LIMIT', 15 * 10**6))

# Deprecated: FLASHBOTS_RPC URL
# Mainnet: "https://relay.flashbots.net",
# Görli: "https://relay-goerli.flashbots.net",
FLASHBOTS_RPC = os.getenv('FLASHBOTS_RPC', '')
AUCTION_BUNDLER_URIS = os.getenv('AUCTION_BUNDLER_URIS', FLASHBOTS_RPC).split(',')

# Deprecated: FLASHBOT_SIGNATURE
FLASHBOT_SIGNATURE = os.getenv('FLASHBOT_SIGNATURE', None)
AUCTION_BUNDLER_PRIVATE_KEY = os.getenv('AUCTION_BUNDLER_PRIVATE_KEY', FLASHBOT_SIGNATURE)

# Curated module strategy
GAS_FEE_PERCENTILE_1: int = int(os.getenv('GAS_FEE_PERCENTILE_1', 20))
GAS_FEE_PERCENTILE_DAYS_HISTORY_1: int = int(os.getenv('GAS_FEE_PERCENTILE_DAYS_HISTORY_1', 1))

GAS_PRIORITY_FEE_PERCENTILE = int(os.getenv('GAS_PRIORITY_FEE_PERCENTILE', 25))

MAX_BUFFERED_ETHERS = Web3.to_wei(*os.getenv('MAX_BUFFERED_ETHERS', '5000 ether').split(' '))

# Metrics
PROMETHEUS_PORT = int(os.getenv('PROMETHEUS_PORT', '9000'))
PULSE_SERVER_PORT = int(os.getenv('PULSE_SERVER_PORT', '9010'))
MAX_CYCLE_LIFETIME_IN_SECONDS = int(os.getenv('MAX_CYCLE_LIFETIME_IN_SECONDS', '1200'))

# List of ids of staking modules in which the depositor bot will make deposits
DEPOSIT_MODULES_WHITELIST = [int(module_id) for module_id in os.getenv('DEPOSIT_MODULES_WHITELIST', '1').split(',')]
